c10/core/TensorImpl.h

```diff
@@ -138,6 +138,68 @@ struct C10_API AutogradMetaInterface {
   virtual ~AutogradMetaInterface();
 };
 
+// NOTE [ Version Counter Sharing ]
+//
+// Every Tensor has a version counter. Version counters are incremented
+// whenever the data or shape of a tensor changes through Variable operations.
+// These are typically in-place operations. Version counters are used to
+// detect modifications to saved variables which would result in incorrect
+// gradient calculations. Version counters may be shared between Variables:
+//
+// 1. A view shares the version counter of the base Variable,
+// 2. Detached variables share the version counter of the source,
+// 3. Unpacked saved variables share the version counter of the source.
+//
+// Version counters are not shared when:
+//
+// 1. We replace a `Variable`'s underlying `Tensor` by calling `set_data(...)`.
+//
+// Question 1: Why do we not increment the version counter in non-Variable
+// operations?
+//
+// Answer: We explicitly don't increment the version counter in non-Variable
+// operations, because it is a useful escape hatch when we want to change the
+// content of a Variable without invalidating it in the autograd graph.
+//
+// Question 2: Why do we put the version counter in TensorImpl instead of AutogradMeta?
+//
+// Answer: After the Variable/Tensor merge, a tensor will not have AutogradMeta when
+// its `requires_grad_` is false, but when we use this tensor in the forward pass of
+// a function that requires saving this tensor for backward, we need to keep track of
+// this tensor's version to make sure it's always valid in the autograd graph.
+//
+// One logical way to achieve this goal is to initialize AutogradMeta and create the
```

Comment:
I'd make it clearer here that you are talking about a *hypothetical* alternative way to work around this problem, but not what is actually implemented.